api4/cloud_test.go

```diff
@@ -141,3 +141,91 @@ func Test_getCloudLimits(t *testing.T) {
 		require.Equal(t, *mockLimits.Messages.History, *limits.Messages.History)
 	})
 }
+
+func TestGetCloudUsageForMessages(t *testing.T) {
+	t.Run("no license returns not implemented", func(t *testing.T) {
+		th := Setup(t)
+		defer th.TearDown()
+
+		th.App.Srv().RemoveLicense()
+
+		usage, r, err := th.Client.GetCloudUsageForMessages()
+		require.Error(t, err)
+		require.Nil(t, usage)
+		require.Equal(t, http.StatusNotImplemented, r.StatusCode)
+	})
+
+	t.Run("non cloud license returns not implemented", func(t *testing.T) {
+		th := Setup(t)
+		defer th.TearDown()
+
+		th.App.Srv().SetLicense(model.NewTestLicense())
+
+		usage, r, err := th.Client.GetCloudUsageForMessages()
+		require.Error(t, err)
+		require.Nil(t, usage)
+		require.Equal(t, http.StatusNotImplemented, r.StatusCode)
+	})
+
+	t.Run("feature flag off returns empty usage", func(t *testing.T) {
+		th := Setup(t)
+		defer th.TearDown()
+
+		os.Setenv("MM_FEATUREFLAGS_CLOUDFREE", "false")
+		defer os.Unsetenv("MM_FEATUREFLAGS_CLOUDFREE")
+		th.App.ReloadConfig()
+
+		th.App.Srv().SetLicense(model.NewTestLicense("cloud"))
+
+		usage, r, err := th.Client.GetCloudUsageForMessages()
+		require.NoError(t, err)
+		require.Equal(t, &model.MessagesUsage{}, usage)
+		require.Equal(t, http.StatusOK, r.StatusCode)
+	})
+
+	t.Run("unauthenticated users can not access", func(t *testing.T) {
+		th := Setup(t)
+		defer th.TearDown()
+
+		th.Client.Logout()
+
+		usage, r, err := th.Client.GetCloudUsageForMessages()
+		require.Error(t, err)
+		require.Nil(t, usage)
+		require.Equal(t, http.StatusUnauthorized, r.StatusCode)
+	})
+
+	t.Run("good request with cloud server and feature flag returns response", func(t *testing.T) {
+		th := Setup(t)
+		defer th.TearDown()
+
+		os.Setenv("MM_FEATUREFLAGS_CLOUDFREE", "true")
+		defer os.Unsetenv("MM_FEATUREFLAGS_CLOUDFREE")
+		th.App.ReloadConfig()
+		th.App.Srv().SetLicense(model.NewTestLicense("cloud"))
+
+		cloud := &mocks.CloudInterface{}
+		max := 10
+		mockLimits := &model.ProductLimits{
+			Messages: &model.MessagesLimits{
+				History: &max,
+			},
+		}
+		cloud.On("GetCloudLimits", mock.Anything).Return(mockLimits, nil)
+		th.App.Srv().Cloud = cloud
+
+		// The following three calls create total of 4 posts
+		th.BasicTeam = th.CreateTeam()
+		th.BasicChannel = th.CreatePublicChannel()
+		th.BasicPost = th.CreatePost()
```

Comment:
good call! no, system messages shouldn't count. But I don't think it will matter much in the first months. Can we create a ticket to adress it properly later on? right now we are not enforcing, but once we enforce them people is likely to complain.